pkg/kubelet/kuberuntime/fake_kuberuntime.go

```diff
@@ -0,0 +1,554 @@
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "DeletePodSandbox")
+
+	if _, ok := r.Sandboxes[podSandBoxID]; ok {
+		delete(r.Sandboxes, podSandBoxID)
+	}
+
+	return nil
+}
+
+func (r *fakeKubeRuntime) PodSandboxStatus(podSandBoxID string) (*runtimeApi.PodSandboxStatus, error) {
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "PodSandboxStatus")
+
+	s, ok := r.Sandboxes[podSandBoxID]
+	if !ok {
+		return nil, fmt.Errorf("PodSandbox not found")
+	}
+
+	return &runtimeApi.PodSandboxStatus{
+		Id:        &podSandBoxID,
+		Name:      s.config.Name,
+		CreatedAt: &s.createdAt,
+		State:     &s.state,
+		Network: &runtimeApi.PodSandboxNetworkStatus{
+			Ip: &fakePodSandboxIP,
+		},
+		Labels:      s.config.Labels,
+		Annotations: s.config.Annotations,
+	}, nil
+}
+
+func filterInLabels(filter, labels map[string]string) bool {
+	for k, v := range filter {
+		if value, ok := labels[k]; ok {
+			if value != v {
+				return false
+			}
+		} else {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (r *fakeKubeRuntime) ListPodSandbox(filter *runtimeApi.PodSandboxFilter) ([]*runtimeApi.PodSandbox, error) {
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "ListPodSandbox")
+
+	result := make([]*runtimeApi.PodSandbox, 0)
+	for id, s := range r.Sandboxes {
+		if filter != nil {
+			if filter.Id != nil && filter.GetId() != id {
+				continue
+			}
+			if filter.Name != nil && filter.GetName() != s.config.GetName() {
+				continue
+			}
+			if filter.State != nil && filter.GetState() != s.state {
+				continue
+			}
+			if filter.LabelSelector != nil && !filterInLabels(filter.LabelSelector, s.config.Labels) {
+				continue
+			}
+		}
+
+		result = append(result, &runtimeApi.PodSandbox{
+			Id:        &id,
+			Name:      s.config.Name,
+			State:     &s.state,
+			CreatedAt: &s.createdAt,
+			Labels:    s.config.Labels,
+		})
+	}
+
+	return result, nil
+}
+
+func (r *fakeKubeRuntime) CreateContainer(podSandBoxID string, config *runtimeApi.ContainerConfig, sandboxConfig *runtimeApi.PodSandboxConfig) (string, error) {
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "CreateContainer")
+
+	imageID := ""
+	for _, img := range r.Images {
+		if stringInSlice(config.Image.GetImage(), img.RepoTags) {
+			imageID = img.GetId()
+			break
+		}
+	}
+	if imageID == "" {
+		return "", fmt.Errorf("Image %s not found", config.Image.GetImage())
+	}
+
+	containerID := config.GetName()
+	r.Containers[containerID] = &ContainerWithState{
+		createdAt:       time.Now().Unix(),
+		containerID:     containerID,
+		podSandBoxID:    podSandBoxID,
+		containerConfig: config,
+		sandboxConfig:   sandboxConfig,
+		imageID:         imageID,
+		state:           runtimeApi.ContainerState_CREATED,
+	}
+
+	return containerID, nil
+}
+
+func (r *fakeKubeRuntime) StartContainer(rawContainerID string) error {
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "StartContainer")
+
+	if c, ok := r.Containers[rawContainerID]; ok {
```

Comment:
nit: it might be slightly more clear to handle the error case first and return. E.g.,
```
c, ok := r.Containers[rawContainerID]
if !ok {
    return fmt.Errorf("container not found")
}
// Go on with the normal routine
```
This is my personal preference to reduce the if/else blocks.